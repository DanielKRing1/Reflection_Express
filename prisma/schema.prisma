// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL_LOCAL")
}

model User {
  password_userId             Password?           @relation(name: "password_userId")
  journal_userId              Journal[]           @relation(name: "journal_userId")
  email                       String              @id @unique

  // Set to null when delete Journal + Update id if Journal id updated
  user_lastUsedJournalId      Journal?            @relation(name: "user_lastUsedJournalId", fields: [lastUsedJId], references: [id], onDelete: SetNull, onUpdate: Cascade)
  lastUsedJId                 BigInt?             @unique
}

model Password {
  // Cascade delete/update with User
  password_userId             User                @relation(name: "password_userId", fields: [userId], references: [email], onDelete: Cascade, onUpdate: Cascade)
  userId                      String              @id

  hash                        String
}

model Journal {
  user_lastUsedJournalId      User?               @relation(name: "user_lastUsedJournalId")
  inkling_journalId           Inkling[]           @relation(name: "inkling_journalId")
  journalentry_journalId      JournalEntry[]      @relation(name: "journalentry_journalId")
  thought_journalId           Thought[]           @relation(name: "thought_journalId")
  reflection_journalId        Reflection[]        @relation(name: "reflection_journalId")
  id                          BigInt              @id @default(autoincrement())
  
  // Cascade delete/update with User
  journal_userId              User                @relation(name: "journal_userId", fields: [userId], references: [email], onDelete: Cascade, onUpdate: Cascade)
  userId                      String

  @@index([userId])
}

model Inkling {
  timeId                      DateTime            @db.Timestamptz(3)

  // Cascade delete/update with Journal
  inkling_journalId           Journal             @relation(name: "inkling_journalId", fields: [journalId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  journalId                   BigInt

  text                        String

  @@id([journalId, timeId])
}

model JournalEntry {
  reflection_journalEntryId   Reflection[]        @relation(name: "reflection_journalEntryId")
  timeId                      DateTime            @db.Timestamptz(3)

  // Cascade delete/update with Journal
  journalentry_journalId      Journal             @relation(name: "journalentry_journalId", fields: [journalId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  journalId                   BigInt

  @@id([journalId, timeId])
}

model Thought {
  reflection_thoughtId        Reflection[]        @relation(name: "reflection_thoughtId")
  timeId                      DateTime            @db.Timestamptz(3)

  // Cascade delete/update with Journal
  thought_journalId           Journal             @relation(name: "thought_journalId", fields: [journalId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  journalId                   BigInt

  text                        String

  @@id([journalId, timeId])
}

model Reflection {
  // Cascade delete/update with Journal
  reflection_journalId        Journal             @relation(name: "reflection_journalId", fields: [journalId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  journalId                   BigInt

  // Cascade delete/update with JournalEntry
  reflection_journalEntryId   JournalEntry        @relation(name: "reflection_journalEntryId", fields: [journalId, journalEntryId], references: [journalId, timeId], onDelete: Cascade, onUpdate: Cascade)
  journalEntryId              DateTime            @db.Timestamptz(3)

  // Cascade delete/update with Thought
  reflection_thoughtId        Thought             @relation(name: "reflection_thoughtId", fields: [journalId, thoughtId], references: [journalId, timeId], onDelete: Cascade, onUpdate: Cascade)
  thoughtId                   DateTime            @db.Timestamptz(3)

  decision                    Int                 @db.SmallInt

  @@id([journalEntryId, thoughtId])
}
