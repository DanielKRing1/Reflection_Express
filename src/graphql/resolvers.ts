import DateScalarFragment from "./scalars/date.gql";

import UserSchemaFragment from "./routes/user/resolvers.gql";
import InklingSchemaFragment from "./routes/inkling/resolvers.gql";
import JournalEntrySchemaFragment from "./routes/journalEntry/resolvers.gql";
import JournalSchemaFragment from "./routes/journal/resolvers.gql";

export default {
    ...DateScalarFragment.Resolver,

    Query: {
        ...UserSchemaFragment.Query,
        ...InklingSchemaFragment.Query,
        ...JournalEntrySchemaFragment.Query,
        ...JournalSchemaFragment.Query,
    },

    Mutation: {
        ...UserSchemaFragment.Mutation,
        ...InklingSchemaFragment.Mutation,
        ...JournalEntrySchemaFragment.Mutation,
        ...JournalSchemaFragment.Mutation,
    },
};
