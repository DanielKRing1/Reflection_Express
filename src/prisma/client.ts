import { PrismaClient } from "./generated/client";

let prisma = new PrismaClient({
    // log: ["query"],
});

export default prisma;
