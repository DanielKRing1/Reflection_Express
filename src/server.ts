// EXPRESS
import express, { Express } from "express";
import cors from "cors";

// MIDDLEWARE
import loggingMiddleware from "./middlewares/logging.middleware";

// ROUTES
import genRouter from "./routes";

// CONFIG
import { PORT } from "./config/server.config";

// APP
const app: Express = express();

// MIDDLEWARE
app.use(cors());
// parse requests of content-type - application/json
app.use(express.json());
// parse requests of content-type - application/x-www-form-urlencoded
app.use(
  express.urlencoded({
    extended: true,
  })
);
app.use(loggingMiddleware);

// GRAPHQL
app.use(genRouter());

// START SERVER
app.listen(PORT, () => console.log(`Server listening on port ${PORT}!`));
