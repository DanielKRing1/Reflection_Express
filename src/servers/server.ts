import genExpressApp from "./app";
import gqlServer from "./gqlServer";

import accessPromise from "../middlewares/session/access";

export default (async () => {
    const app = await genExpressApp();
    const access = await accessPromise();

    return gqlServer(app, access);
})();
