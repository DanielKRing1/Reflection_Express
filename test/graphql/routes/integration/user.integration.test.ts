import http from "http";
import supertest from "supertest";

import { ApolloServer } from "@apollo/server";

import server from "../../../../src/index";
import typeDefs from "../../../../src/graphql/schema";
import resolvers from "../../../../src/graphql/resolvers";
import GqlContext from "../../../../src/graphql/types/context.types";
import prisma from "../../../../src/prisma/client";

const USER_ID = "1";
const PASSWORD = "123";

const gqlServer = new ApolloServer<GqlContext>({
    typeDefs,
    resolvers,
});

describe("User GraphQL api", () => {
    let s: http.Server;
    let request: supertest.SuperTest<supertest.Test>;

    beforeAll(async () => {
        await prisma.$queryRaw`TRUNCATE TABLE "User" CASCADE;`;

        // TEST EXPRESS SESSION ROUTES
        s = await server;
        request = supertest(s);

        await request
            .post("/login/create-user")
            .send({ userId: USER_ID, password: PASSWORD });
    });

    it("Should get the created User", async () => {
        // run the query against the server and snapshot the output
        const res = await gqlServer.executeOperation(
            {
                query: `query User {
                    user {
                        email
                        lastUsedJId
                    }
                }`,
                variables: {},
            },
            {
                contextValue: {
                    // @ts-ignore
                    req: { session: { userId: USER_ID } } as unknown as Request,
                    // @ts-ignore
                    res: {},
                },
            }
        );

        // console.log(res);
        // // @ts-ignore
        // console.log(res.body.singleResult.data);
        // console.log(JSON.stringify(res.body));

        expect(res).toMatchSnapshot();
    });

    afterAll(async () => {
        await prisma.$queryRaw`TRUNCATE TABLE "User" CASCADE;`;

        await prisma.$disconnect();

        await s.close();
    });
});
