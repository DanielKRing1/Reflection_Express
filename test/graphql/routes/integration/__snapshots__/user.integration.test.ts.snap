// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User GraphQL api Should get the created User 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "user": {
          "email": "1",
          "lastUsedJId": null,
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;
